bip-testnet4.mediawiki

```diff
@@ -0,0 +1,102 @@
+<pre>
+  BIP: ?
+  Layer: Applications
+  Title: Testnet 4
+  Author: Fabian Jahr <fjahr@protonmail.com>
+  Comments-Summary: No comments yet.
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-testnet4
+  Status: Draft
+  Type: Standards Track
+  Created: 2024-05-27
+  License: CC0-1.0
+  Post-History: https://groups.google.com/g/bitcoindev/c/9bL00vRj7OU/m/9yCPo3uUBwAJ
+                https://github.com/bitcoin/bitcoin/pull/29775
+</pre>
+
+== Abstract ==
+
+A new test network with the goal to replace Testnet 3. This network comes with small but important improvements of the consensus rules, that should make it impractical to attack the network using only CPU mining.
+
+== Motivation ==
+
+Quoting the original mailing list post from Jameson Lopp:
+
+"Testnet3 has been running for 13 years. It's on block 2.5 million something and the block reward is down to ~0.014 TBTC, so mining is not doing a great job at distributing testnet coins anymore.
+
+The reason the block height is insanely high is due to a rather amusing edge case bug that causes the difficulty to regularly get reset to 1, which causes a bit of havoc. If you want a deep dive into the quirk: https://blog.lopp.net/the-block-storms-of-bitcoins-testnet/
+
+Testnet3 is being actively used for scammy airdrops; those of us who tend to be generous with our testnet coins are getting hounded by non-developers chasing cheap gains.
+
+As a result, TBTC is being actively bought and sold; one could argue that the fundamental principle of testnet coins having no value has been broken."
+
+Since then the issue with block storms has been further demonstrated on Testnet 3 when three years' worth of blocks were mined in a few weeks while rendering the network practically unusable at the same time.
+
+== Specification ==
+
+Consensus of Testnet 4 follows the same rules as Testnet 3 with the exception of the two new rules detailed below.
+
+This means that the existing 20 min difficulty exception in Testnet 3 is explicitly kept in place, meaning that a block with a timestamp 20 minutes further into the future from the current time is allowed to have a minimum difficulty of 1 instead of whatever the actual level currently is.
+
+=== Block Storm Fix ===
+
+When the next work required is calculated for the first block in a new difficulty period, the last block of the previous difficulty period is only used as the base for this calculation if its difficulty is not 1. If its difficulty is 1, all blocks in the previous difficulty period are checked in reverse order if any of them have a different difficulty than 1. When a different difficulty is encountered, it is assumed to be the actual difficulty of the network and used as the base for the calculation of the new difficulty level. Only if all blocks from the previous difficulty period have had a difficulty of 1, this is also used as the base for the calculation of the next difficulty period.
```

Comment:
> This seems a bit imprecise. It sounds like the latest block is ignored, but I suspect that the _timestamps_ from first and last block are always used, it is just the _difficulty_ that is taken from the latest block with a difficulty greater than 1.

Right, good catch. I tried to clarify this in this last edit.

> Perhaps you could also clarify: When every single block in a difficulty period is mined per the 20-minute exception, the difficulty would still reset to 1, right?

In theory but that isn't possible. The first block of the difficulty period does not allow the usage of the 20-min rule. I managed to confuse myself with this again as well so I changed this further and clarified this. The confusing part here is that this is not a change of behavior which is why I failed to emphasize this. The first block of the difficulty period also can not use the 20-min rule in Testnet 3. But there it can get difficulty 1 when the last block of the previous period was mined with the 20-min rule, which then leads to a block storm.

Reply:
Oh, but then that means that if someone drives up the difficulty by mining a few difficulty periods with an ASIC and then leaves right after mining the last block of a difficulty period, the network does get stuck on the first block of the next difficulty period, because it does have to be mined at actual difficulty.

Reply:
Yeah, that has been discussed in the PR, let's continue the discussion there :)